src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -562,6 +566,635 @@ def _encloses_omp_directive(self):
+
+        # TODO Generate extra clauses
+        # Find the parent Loop node
+        parent_loop = self.ancestor(Loop)
+
+        # Find the parent OMPParallelDirective
+        parallel_directive = self.ancestor(OMPParallelDirective)
+        parallel_private = parallel_directive._get_private_list()
+
+        node = self.children[0].children[0]
+
+        # Find our loop initialisation, variable and bounds
+        loop_var = node.variable
+        start_val = node.children[0]
+        stop_val = node.children[1]
+        step_val = node.children[2]
+        schedule = node.children[3]
+
+        # FIXME Disallow transformation if init/stop/step include array references
+        # FIXME Handle reduction variables
+
+        # Loop variable is private
+        private_list.append(loop_var.name)
+
+        # Any variables used in start_val, stop_val or step_val are firstprivate
+        # Get the References used for initialisation
+        start_val_refs = start_val.walk(Reference)
+        # For all non-array accesses we make them firstprivate
+
+        # TODO Do we worry about array references here? For now we don't allow
+        # array references or ArrayOfStruct references
+        for ref in start_val_refs:
+            if isinstance(ref, (ArrayReference, ArrayOfStructuresReference)):
+                raise GenerationError("{0} not yet supported in the start "
+                                      "variable of the primary Loop in a "
+                                      "OMPTaskDirective node.".format(
+                    type(ref).__name__))
+            name = writer(ref)
+            if ref not in firstprivate_list:
+                firstprivate_list.append(name)
+
+        stop_val_refs = stop_val.walk(Reference)
+        for sig in stop_val_refs:
+            if isinstance(ref, (ArrayReference, ArrayOfStructuresReference)):
+                raise GenerationError("{0} not yet supported in the stop "
+                                      "variable of the primary Loop in a "
+                                      "OMPTaskDirective node.".format(
+                    type(ref).__name__))
+            name = writer(ref)
+            if ref not in firstprivate_list:
+                firstprivate_list.append(name)
+
+        
+        step_val_refs = step_val.walk(Reference)
+        for sig in step_val_refs:
+            if isinstance(ref, (ArrayReference, ArrayOfStructuresReference)):
+                raise GenerationError("{0} not yet supported in the step "
+                                      "variable of the primary Loop in a "
+                                      "OMPTaskDirective node.".format(
+                    type(ref).__name__))
+            name = writer(ref)
+            if ref not in firstprivate_list:
+                firstprivate_list.append(name)
+
+        # Look through the schedule and work out data sharing clauses.
+        statements = schedule.walk((Assignment, IfBlock, Loop))
```

Comment:
Could this be implemented using a Visitor instead?

Reply:
Yes I think so - but I'm not sure how to do this for PSyclone directives/where to put the visitor in the codebase. Is there an example I can look at?